src/DotNetty.Common/Internal/AppendableCharSequence.cs

```diff
@@ -0,0 +1,219 @@
+            ? StringComparer.OrdinalIgnoreCase.GetHashCode(this.ToString())
+            : StringComparer.Ordinal.GetHashCode(this.ToString());
+
+        public override int GetHashCode() => this.HashCode(true);
+
+        public bool SequenceEquals(ICharSequence other, bool ignoreCase) =>
+            CharUtil.SequenceEquals(this, other, ignoreCase);
+
+        public IAppendable Append(char c)
+        {
+            try
+            {
+                this.chars[this.pos++] = c;
+            }
+            catch (IndexOutOfRangeException)
+            {
+                this.Expand();
+                this.chars[this.pos - 1] = c;
+            }
+
+            return this;
+        }
+
+        public IAppendable Append(ICharSequence sequence) => this.Append(sequence, 0, sequence.Count);
+
+        public IAppendable Append(ICharSequence sequence, int start, int end)
+        {
+            Contract.Requires(sequence.Count >= end);
+
+            int length = end - start;
+            if (length > this.chars.Length - this.pos)
+            {
+                this.chars = Expand(this.chars, this.pos + length, this.pos);
+            }
+
+            var seq = sequence as AppendableCharSequence;
+            if (seq != null)
+            {
+                // Optimize append operations via array copy
+                char[] src = seq.chars;
+                Array.Copy(src, start, this.chars, this.pos, length);
+                this.pos += length;
+
+                return this;
+            }
+
+            for (int i = start; i < end; i++)
+            {
+                this.chars[this.pos++] = sequence[i];
+            }
+
+            return this;
+        }
+
+        /**
+          * Reset the {@link AppendableCharSequence}. Be aware this will only reset the current internal position and not
+          * shrink the internal char array.
+          */
+        public void Reset() => this.pos = 0;
+
+        public string ToString(int start)
+        {
+            Contract.Requires(start >= 0 && start < this.Count);
+
+            return new string(this.chars, start, this.pos);
+        }
+
+        public override string ToString() => this.Count == 0 ? string.Empty : this.ToString(0);
+
+        /**
+          * Create a new {@link String} from the given start to end.
+          * This method is considered unsafe as index values are assumed to be legitimate.
+          * Only underlying array bounds checking is done.
+         */
+        public string SubStringUnsafe(int start, int end) => new string(this.chars, start, end - start);
+
+        void Expand()
+        {
+            char[] old = this.chars;
+            // double it
+            int len = old.Length << 1;
+            if (len < 0)
+            {
+                throw new InvalidOperationException($"Length {len} must be positive");
+            }
+
+            this.chars = new char[len];
+            Buffer.BlockCopy(old, 0, this.chars, 0, old.Length);
```

Comment:
`Buffer.BlockCopy(old, 0/* * sizeof(char)*/, this.chars, 0/* * sizeof(char)*/, old.Length * sizeof(char));` or `Array.Copy`